### ISSUE TITLE
v3.0.0 Release Tracker

### ISSUE THREAD

**Issue by redonkulus** [formatjs/react-intl#1280]
@DragonRaider5 

Creating this issue to track the remaining PR's/Issues for the v3.0.0 release. I've created a new [milestone](https://github.com/formatjs/react-intl/milestone/2) to make it easy to see the PR's that are open.

Can you help fill in the [v3.0.0 Migration guide](https://github.com/formatjs/react-intl/wiki/Upgrade-Guide#300)? It would be good to follow the format from the v2 section.

**Comment by leoyli** [comment]
Does anything need to submit to DefinitelyTyped for TS user?

**Comment by okuryu** [comment]
It seems that the related issue has already been filed.
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/35812

**Comment by redonkulus** [comment]
What’s the turnaround time on getting PRs merged in that repo? I assume we can release our changes independent of their release cycle?

**Comment by leoyli** [comment]
@redonkulus, I think so too since this project is not obligated to maintain the type declarations. But it would be nice to see it can come together, so just rase the attention (and cross reference issues).

**Comment by redonkulus** [comment]
As as of right now, the only remaining tasking before publishing is to finish the Migration guide. @DragonRaider5 do you have an ETA of when you think this can be done?

If there are no other breaking changes we would like to make, I believe we can let the current beta sit for a few days, let users try it out and then publish v3.0.0.

**Comment by leoyli** [comment]
Does anyone have a clue on why the bundle size increased significantly?
https://bundlephobia.com/result?p=react-intl@3.0.0-beta-6

![size](https://user-images.githubusercontent.com/8866555/58841516-36f7a280-8638-11e9-8db0-4868cd8df12c.png)

**Comment by redonkulus** [comment]
@longlho anything related to your recent downstream changes?

@leoyli can you check against react-intl@v2? maybe we can get a more detailed comparison of the downstream dependencies.

**Comment by longlho** [comment]
Yeah it's prob bc of the polyfill for `Intl.PluralRules`. `make-plural` isn't new, it was just bundled in with the old version of `intl-messageformat` (sans the CLDR data).

**Comment by leoyli** [comment]
This part I have no clue. It seems people are really care the bundling size for some use case and that is also part of benefit this package now drop React 0.14 skip 15. But now this benefit seems not able to be justified... Maybe it's okay or someone have a better idea on how to down size it. 🤔

**Comment by DragonRaider5** [comment]
What about #1293? `1.3kb` to `1.43kb` is a considerable change, as that's about a 10% increase and might explain the explosion.

However - does bundlephobia take into account conditional dependencies?
Users won't just load all the locale data.

**Comment by longlho** [comment]
I think this is resolved :) beta-7 is significantly smaller than previous
releases now

**Comment by leoyli** [comment]
Yes! My concerns was resolved and now looks super awesome, thank you for the great work @longlho ❤️

![size](https://user-images.githubusercontent.com/8866555/59109564-419d8a80-890b-11e9-8ea5-f0659c9a8474.png)

**Comment by leoyli** [comment]
@redonkulus, since the last changes at the code level (beta.6) is out for almost 1.5 weeks.  Do you have a targeted date in mind?! Tx!

**Comment by redonkulus** [comment]
@leoyli I published beta.8 which includes some optimizations that @longlho introduced. We should let that marinate for a little bit to see if there are any performance issues.

**Comment by redonkulus** [comment]
One last change we are waiting on is the new FormattedRelativeTime changes that @longlho is working on. It’s important to get this merged before we release 3.x.

**Comment by longlho** [comment]
I think last step is to update the examples and make sure they still work

**Comment by leoyli** [comment]
Should we strip the `src` and `lib` folder in the tarball? For the enduser, they not really care about the source code much.  Not just to squeeze the bandwidth, but on the aspects of DX. It is very annoying that WebStorm suggest me to use something coming from the `src`, `lib`...

![Screen Shot 2019-07-12 at 11 17 31 PM](https://user-images.githubusercontent.com/8866555/61166436-5fa66e00-a4fb-11e9-8677-3b2ad7c1d743.png)

**Comment by longlho** [comment]
We can ignore src yeah. Lib has esm version

**Comment by leoyli** [comment]
Oh, so is that means if we were used Webpack, then we probably want to import methods from `lib` instead.  I will assume that means we can do tree shaking?!

**Comment by longlho** [comment]
Yes

**Comment by leoyli** [comment]
@longlho, could we do the other way around? I mean to let the top-level expose ESM, and `/lib` for CJS.  Since most people now use react with webpack (or other build tool), I think to support ESM is much sounding, and even the latest NodeJS supports ESM now.

Also, I think this part need to be documented in the upgrade guide.

**Comment by leoyli** [comment]
@longlho, could we do the other way around? I mean to let the top-level expose ESM, and /lib for CJS. Since most people now use react with webpack (or other build tools), I think to support ESM out of the box is much sounding (I don't know why we still need CJS in the frontend world...), and even the latest NodeJS supports ESM now. This explains why my ID ask me to import file from the lib folder... 😓

Also, I think this part need to be documented in the upgrade guide. 🤔

**Comment by longlho** [comment]
Let me give this some thoughts, we can actually dist out `.mjs` for Node 12.

**Comment by fletcher91** [comment]
@longlho Depending on project setup, setting `type: "module"` in package.json [lets Node detect es modules with the regular .js extensions](https://medium.com/@nodejs/announcing-a-new-experimental-modules-1be8d2d6c2ff#b023)

**Comment by longlho** [comment]
yeah that works as well. I might just drop CJS like @leoyli suggested & set `type: module`

**Comment by YUFENGWANG** [comment]
Is version 3.0 ready?

**Comment by longlho** [comment]
Almost, #513 is the last issue I'm trying to tackle before official v3

**Comment by iamstarkov** [comment]
How open are you to accept a PR with the implementation of rest of the hooks?

**Comment by longlho** [comment]
@iamstarkov can u clarify `rest of the hooks`?

**Comment by iamstarkov** [comment]
right now, there is only `useIntl` and thats it, I maybe need rest of the components as hooks

**Comment by longlho** [comment]
Right now you can do `const {intl: {formatMessage}} = useIntl()`. Is that not sufficient? What's the API you're looking for?

**Comment by iamstarkov** [comment]
That’s seems fine

**Comment by leoyli** [comment]
Side notes:

> To keep the API surface clean and simple, we only provide useIntl hook in the package. If preferable, user can wrap this built-in hook to make customized hook like useFormatMessage easily. Please visit React's official website for more general introduction on React hooks.

https://github.com/formatjs/react-intl/blob/master/docs/Upgrade-Guide.md#new-useintl-hook-as-an-alternative-of-injectintl-hoc

**Comment by paolostyle** [comment]
Any chance you'd provide a non-hacky way to use internationalization outside of React components (e.g. in redux-saga) in v3? Or at least a documented way how to do that?

**Comment by longlho** [comment]
@paolostyle 	can u create a separate issue to track? Also include some details about your use case please?

**Comment by leoyli** [comment]
@redonkulus, @longlho, @DragonRaider5, dear maintainers,

Thank you for putting much love here to keep this library evergreen.  Highly appreciate that 👍 If possible, could any of you explains a bit the actual plan and thinks to be tracked?!  I don't see it much here but I do see new features keeping added but not really planned.  Is that possible to delay some of that into 3.1?

I think the most important one that people have kept waiting too long is fixing the legacy react context API so it works on the `StrictMode`. This have been fixed for a while.  And we have also dropped many polyfills and add hook support even with Rich text that just added couple days ago. And today I see a new beta adding the ability to use this lib out of the React's realm, which is really awesome and useful for people using SSR.

All of these are great but I think the release cycle is too long.  Some problems can only be tested by greater adaption only if that can be officially released. So I hope to know when we can really drop the `beta` flag so we can use it on production.  Like I really want to use hook but some of my co-worker are really conservative, even it work perfectly fine...

Thank you :)

**Comment by longlho** [comment]
I think we're done w/ breaking changes (the last PR did that). I'm letting it bake over the weekend and am testing internally within Dropbox. v3 will drop next Wednesday.

**Comment by leoyli** [comment]
That’s a great news! I totally understand we need it to be baked. Just need a rough timeline so people can start to plan their tasks, so don’t feel too much pressure. Good meal deserves it’s baking time, even sometimes it can be long!

**Comment by longlho** [comment]
If nothing horrible happens, next Tuesday/Wednesday :) After that 3.1/3.2 hopefully will have ICU Skeleton support for [datetime](http://userguide.icu-project.org/formatparse/datetime)(e.g: `today is {ts, date, ::YYYY-MM-ss}`) & [number](http://site.icu-project.org/design/formatting/numbers/skeleton) (e.g: `number is {num, number, ::0.##}`)

**Comment by longlho** [comment]
v3 has been released 🎉 https://github.com/formatjs/react-intl/releases/tag/v3.0.0 thanks a lot everyone for your contributions!!!

**Comment by YUFENGWANG** [comment]
Bravo!

**Comment by longlho** [comment]
I've opened a new GH issue to track 3.1 https://github.com/formatjs/react-intl/issues/1392

**Comment by iamstarkov** [comment]
hurray
